src/liballoc/raw_vec.rs

```diff
@@ -350,19 +351,25 @@ impl<T> RawVec<T> {
 
             // NOTE: we don't early branch on ZSTs here because we want this
             // to actually catch "asking for more than usize::MAX" in that case.
-            // If we make it past the first branch then we are guaranteed to
-            // panic.
+            let required_cap = used_cap.checked_add(needed_extra_cap)
+               .expect("capacity overflow");
```

Comment:
We must move this overflow check and panic case to below the `required_cap <= self.cap` check. While preserving this panic if it's a ZST element type.

The reason is that the fast path is the section of code before we decide if we need to reallocate or not, and it should be free from a branch to panic.